### PR TITLE
VEP 165: Containerpath Volumes

### PR DESCRIPTION
**Pull request by mhenriks**
### What this PR does

  Summary

  ContainerPath volumes allow VirtualMachines to access files and directories from the virt-launcher pod's filesystem via virtiofs. This enables VMs to consume data dynamically injected into pods by Kubernetes or platform-specific mechanisms.

  Motivation

  Kubernetes and cloud platforms provide mechanisms for injecting credentials into pods at runtime:
  - Projected service account tokens for workload identity (AWS IRSA, GCP Workload Identity)
  - Dynamic secrets via admission webhooks (Vault agent, cert-manager)

  These mechanisms inject data directly into pod containers, but VMs running inside those pods cannot natively access this injected data. ContainerPath volumes bridge this gap by exposing specific paths from the virt-launcher container to the guest VM.

  Implementation

  - New ContainerPathVolumeSource API type with path field
  - ContainerPathVolumes feature gate (Alpha)
  - Pod mutating webhook injects virtiofs volume mounts into the compute container
  - Virtiofs sidecar container generation is skipped (path already exists in compute container)
  - Path existence validation via MissingVirtiofsContainers condition
  - User documentation with AWS IRSA example

  Test plan

  - Unit tests for validation and controller logic
  - E2e tests for basic ContainerPath functionality
  - E2e tests for webhook-injected volumes (emptyDir and ConfigMap)
  - E2e test for live migration with ContainerPath volumes
 

### References

  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issues/165

### Why we need it and why it was done in this way
The following tradeoffs were made:

virtiofs containers injected in mutating webhook on virt-launcher pods. This seems like the most natural way to implement this feature. This way we can validate that the volumes are actually injected into the compute container. Webhook has failurePolicy: Ignore and will only be instantiated if the featuregate is on

The following alternatives were considered:

Could add virtiofs containers to virt-launcher when virt-controller creates the Pod and pray that they actually get added

Links to places where the discussion took place: 

### Special notes for your reviewer

The new `test-helpers` container contains a test-only webhook (test-pod-mutator) used in e2e tests to simulate external volume injection into virt-launcher pods. Not used in production.

I figure this container could be useful for any testing specific programs in the future

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
VEP 165: Containerpath Volumes
```

### CODE REVIEW

**Comment by ram2valar** [comment]
@mhenriks are we covering these in this PR:

### ⏳ Phase 2 (Future PR)
- VirtioFS volume conversion in virt-launcher
- Runtime path validation:
  - Symlink resolution (e.g., `filepath.EvalSymlinks()`)
  - Verify resolved path is within volumeMounts
  - Safepath-style security checks
- Integration tests (Ginkgo)
- E2E tests with VMs

### ⏳ Phase 3 (Future PR)
- Documentation & examples for AWS/GCP/Azure

as following phase 1 covered in https://github.com/kubevirt/kubevirt/pull/16541. or do you want to club all in this single PR?
### ✅ Phase 1 (THIS PR)
- API type `ContainerPathVolumeSource` with CEL validation
- Path must be absolute (starts with `/`)
- Path must not contain `..` (path traversal protection)
- `readOnly` defaults to `true` 
- `readOnly: false` is rejected via CEL (`self == nil || self == true`)
- Generated code (deepcopy, swagger, CRDs)
- **10 files changed**

**Comment by mhenriks** [comment]
@ram2valar let's handle the entire implementation in this PR

@germag see last commit for symlink handling

**Comment by ram2valar** [comment]
> @ram2valar let's handle the entire implementation in this PR
> 
> @germag see last commit for symlink handling

ok fine in that case will close my PR https://github.com/kubevirt/kubevirt/pull/16541

**Comment by germag** [comment]
not trying to be picky, but validating the path for symlinks and later use that path is racy, it's still vulnerable to a TOCTOU. The correct solution is opening the path while validating, directly using `openat2()` or iterating over the path doing `openat()` (`pkg/safepath` already does the latter), and at the end using the `/proc/sef/fd/FD`.

I understand that you need to allow some symlinks, but I think a strict check can be done after the path is canonicalized and all the symlinks are resolved 


Edit: there is another option (just skimmed the code) since you are already checking the mountId, you can just open the path check if the mountId of the path is equal to the mountId of the mount point and then use the `/proc/sef/fd/FD`

**Comment by mhenriks** [comment]
> not trying to be picky, but validating the path for symlinks and later use that path is racy, it's still vulnerable to a TOCTOU. The correct solution is opening the path while validating, directly using `openat2()` or iterating over the path doing `openat()` (`pkg/safepath` already does the latter), and at the end using the `/proc/sef/fd/FD`.
> 
> I understand that you need to allow some symlinks, but I think a strict check can be done after the path is canonicalized and all the symlinks are resolved
> 
> Edit: there is another option (just skimmed the code) since you are already checking the mountId, you can just open the path check if the mountId of the path is equal to the mountId of the mount point and then use the `/proc/sef/fd/FD`

@germag thanks for the review, this is exactly the information we need. Your thoughts on other commits in this PR would be appreciated as well.

WRT addressing the TOCTOU vulnerability, a few considerations:

1. Fixing this properly would require moving validation into the virtiofsd container - The current validation runs in
  the compute container, but file descriptors don't transfer across containers

2. Error reporting becomes complicated - If validation fails in the virtiofsd container, the error is buried in container logs rather than surfaced as a VMI condition or Kubernetes event. We'd need workarounds like termination messages or shared status files

3. The risk is mainly with user-controlled volumes - For Secrets, ConfigMaps, and Projected volumes (the primary use case), content is controlled by Kubernetes. Exploiting TOCTOU would require privileged access to modify them, which defeats the purpose. PVCs and other persistent volume types are where this is a more realistic concern

Given all that, for now I say we either:

1. Ignore this vulnerability given that an attack is non trivial and would likely need code execution in the pod to race the validation, at which point they probably have other attack vectors available

2. Don't allow containerpath volumes to have subdirs. Then containerpath virtiofs containers are just as safe as existing virtiofs containers

3. Don't support PVCs and other persistent volume types (revisit in beta)

I'm leaning towards 3

Tagging @vladikr as well

**Comment by vladikr** [comment]
@mhenriks IIRC that during the VEP discussion we said that the pr will block all kinds of dangerous paths like (/proc, /sys, /dev, /etc/passwd) but I don't see it here. Can we explicitly block these?

Regarding TOCTOU, if @germag idea is difficult to implement for alpha - I'm +1 for dropping support for pvcs right now. 

Besides that, I like this PR.  
So I'll approve this from my side.

**Comment by germag** [comment]
> > not trying to be picky, but validating the path for symlinks and later use that path is racy, it's still vulnerable to a TOCTOU. The correct solution is opening the path while validating, directly using `openat2()` or iterating over the path doing `openat()` (`pkg/safepath` already does the latter), and at the end using the `/proc/sef/fd/FD`.
> > I understand that you need to allow some symlinks, but I think a strict check can be done after the path is canonicalized and all the symlinks are resolved
> > Edit: there is another option (just skimmed the code) since you are already checking the mountId, you can just open the path check if the mountId of the path is equal to the mountId of the mount point and then use the `/proc/sef/fd/FD`
> 
> @germag thanks for the review, this is exactly the information we need. Your thoughts on other commits in this PR would be appreciated as well.
> 
> WRT addressing the TOCTOU vulnerability, a few considerations:
> 
>     1. Fixing this properly would require moving validation into the virtiofsd container - The current validation runs in
>        the compute container, but file descriptors don't transfer across containers
> 
>     2. Error reporting becomes complicated - If validation fails in the virtiofsd container, the error is buried in container logs rather than surfaced as a VMI condition or Kubernetes event. We'd need workarounds like termination messages or shared status files
> 
>     3. The risk is mainly with user-controlled volumes - For Secrets, ConfigMaps, and Projected volumes (the primary use case), content is controlled by Kubernetes. Exploiting TOCTOU would require privileged access to modify them, which defeats the purpose. PVCs and other persistent volume types are where this is a more realistic concern
> 
> 
> Given all that, for now I say we either:
> 
>     1. Ignore this vulnerability given that an attack is non trivial and would likely need code execution in the pod to race the validation, at which point they probably have other attack vectors available

not so difficult (with PVCs) taking into account that user can mix containers and VM

> 
>     2. Don't allow containerpath volumes to have subdirs. Then containerpath virtiofs containers are just as safe as existing virtiofs containers
> 
>     3. Don't support PVCs and other persistent volume types (revisit in beta)
> 
> 
> I'm leaning towards 3

I completely agree with you, not supporting PVCs/persistent volumes will prevent many issues

> 
> Tagging @vladikr as well

**Comment by mhenriks** [comment]
> @mhenriks IIRC that during the VEP discussion we said that the pr will block all kinds of dangerous paths like (/proc, /sys, /dev, /etc/passwd) but I don't see it here. Can we explicitly block these?
> 
> Regarding TOCTOU, if @germag idea is difficult to implement for alpha - I'm +1 for dropping support for pvcs right now.
> 
> Besides that, I like this PR. So I'll approve this from my side.

Thanks for the review @vladikr!

I updated the PR to to only support kubernetes managed volumes (configmap/secret/etc)

WRT supporting dangerous paths. I don't think that is much of an issue at the moment since we switched from supporting any path in the compute container to only volume mounts. I guess it is possible that someone mounts a configmap to /etc but I bet it would be extremely uncommon and not a big concern?

**Comment by vladikr** [comment]
> > @mhenriks IIRC that during the VEP discussion we said that the pr will block all kinds of dangerous paths like (/proc, /sys, /dev, /etc/passwd) but I don't see it here. Can we explicitly block these?
> > Regarding TOCTOU, if @germag idea is difficult to implement for alpha - I'm +1 for dropping support for pvcs right now.
> > Besides that, I like this PR. So I'll approve this from my side.
> 
> Thanks for the review @vladikr!
> 
> I updated the PR to to only support kubernetes managed volumes (configmap/secret/etc)
> 
> WRT supporting dangerous paths. I don't think that is much of an issue at the moment since we switched from supporting any path in the compute container to only volume mounts. I guess it is possible that someone mounts a configmap to /etc but I bet it would be extremely uncommon and not a big concern?

@mhenriks Thanks! I would prefer us explicitly blocking it. But we can do it in a follow up. @jean-edouard what do you think? 

/approve

File: cmd/test-helpers/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    testonly = True,
+    base = "//:passwd-image",
+    # No entrypoint - let the pod spec define which binary to run
+    # This allows multiple test helper binaries in the same image
+    tars = [
+        ":test-pod-mutator-tar",
+        # Future test helpers can be added here
+    ],
+    visibility = ["//visibility:public"],
+)
```
Comment (mhenriks) [inline cmd/test-helpers/BUILD.bazel:22]: @dhiller @enp0s3 this PR adds a new container image called "test-helpers." It contains a webhook that is used for containerpath volume functests. My justification for this is: 

1. I didn't think it made sense to add this binary to any existing containers (but maybe I'm wrong)
2. This container could be a place to put future functest specific binaries

Make sense to you? Do we have do anything on on quay side to support this? Thanks!

**Comment by germag** [comment]
Recently I have to debug an issue related to a webhook (not related to this PR), so I have a question:

Since this is using a mutating webhook and some cloud providers are also using mutating webhooks to modify the pod, is it possibly to guarantee an order among webhooks?, because I thought it is not. How this works if the cloud provider webhook "runs" before/after this one?

**Comment by mhenriks** [comment]
> Recently I have to debug an issue related to a webhook (not related to this PR), so I have a question:
> 
> Since this is using a mutating webhook and some cloud providers are also using mutating webhooks to modify the pod, is it possibly to guarantee an order among webhooks?, because I thought it is not. How this works if the cloud provider webhook "runs" before/after this one?

@germag `IfNeeded` Reinvocation Policy handles this. Basically the webhook will be invoked again if the object changes

https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#reinvocation-policy

**Comment by germag** [comment]
Ok, I see the `reinvocationPolicy` and all the extra logic to make it idempotent and adding the `volumeMounts`. But   IIRC there is no guarantee to be run again if any additional modification is made, so I still not sure how reliable is this approach, specially if it depends on 3rd parties webhooks not messing it.

**Comment by germag** [comment]
> > Recently I have to debug an issue related to a webhook (not related to this PR), so I have a question:
> > Since this is using a mutating webhook and some cloud providers are also using mutating webhooks to modify the pod, is it possibly to guarantee an order among webhooks?, because I thought it is not. How this works if the cloud provider webhook "runs" before/after this one?
> 
> @germag `IfNeeded` Reinvocation Policy handles this. Basically the webhook will be invoked again if the object changes
> 
> https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#reinvocation-policy

yes, I saw it after posting that comment

**Comment by mhenriks** [comment]
> Ok, I see the `reinvocationPolicy` and all the extra logic to make it idempotent and adding the `volumeMounts`. But IIRC there is no guarantee to be run again if any additional modification is made, so I still not sure how reliable is this approach, specially if it depends on 3rd parties webhooks not messing it.

@germag There could always be a poorly behaving webhook that could screw up the system at any time. So let's talk about ordering and guarantees.

kube-apiserver currently executes up to two rounds of mutating webhooks. In the first round, all webhooks are executed in a specific order (`IfNeeded` webhooks are grouped last). There is a second round if the resource is mutated after an `IfNeeded` webhook already ran. In this round all `IfNeeded` webhooks are executed again in the same order. That's it. An `IfNeeded` webhook is guaranteed to be run a second time if um needed. But things will not devolve into a recursive loop if `IfNeeded` webhooks keep making conflicting updates

Let's take the [aws webhook](https://github.com/aws/amazon-eks-pod-identity-webhook/) for example. It actually doesn't set `IfNeeded` but lets pretend it does. This is the worst that could happen:

1. round 0 at t0: our webhook is called and does nothing
2. round0 at t1: aws webhook injects `aws-token` volume and volumemount for `/var/run/secrets/eks.amazonaws.com/serviceaccount/` in all pod containers
3. round1 at t2: our webhook is called and injects virtiofs container mounting `aws-token` volume at `/virtiofs-data/aws-token`
4. round1 at t3: aws webhook runs again and adds `/var/run/secrets/eks.amazonaws.com/serviceaccount/` mount of `aws-token` to virtiofs container

So there's an unused mount in the virtiofs container. Not a big deal IMO. And won't happen in a real system. I expect most of these types of webhooks to behave similarly (not use `IfNeeded` and not break anything if so)

Even if a webhook goes rogue, virt-controller will detect if a virtiofs container is missing and report it

I think we are good

**Comment by germag** [comment]
not for now, but in future versions, could the creation of the container be removed from the webhook?, I mean, let the virt-controller/services/virtiofs.go create the container as usual sharing the containerPath, and the webhook just need to do the path validation and volume injection, could that be possible?

**Comment by germag** [comment]
> I think we are good

I agree that can be a rare event, but this still assumes order and just another single webhook. I did have issues when more than 2 webhooks are running. But, it's ok most probably this will never happen in this case

**Comment by germag** [comment]
And last, (sorry if you did it, I didn't find it) I think it should block the live migration with empty dirs, I don't think a warning in the docs is enough. 

The guest kernel doesn't handle well that files disappears while in the icache/dcache, I plan to work on it this year to improve the migration reliability.

**Comment by mhenriks** [comment]
> not for now, but in future versions, could the creation of the container be removed from the webhook?, I mean, let the virt-controller/services/virtiofs.go create the container as usual sharing the containerPath, and the webhook just need to do the path validation and volume injection, could that be possible?

So virt-controller adds the container with no mount and _our_  webhook adds the mount? Totally possible now but what do we gain by that?

**Comment by mhenriks** [comment]
> And last, (sorry if you did it, I didn't find it) I think it should block the live migration with empty dirs, I don't think a warning in the docs is enough.
> 
> The guest kernel doesn't handle well that files disappears while in the icache/dcache, I plan to work on it this year to improve the migration reliability.

In cases where emptydir is used I expect the emptydir to be initialized by an initcontainer and optionally be kept up to date with a regular sidecar. This is how kubernetes vault agent works

So assuming the data is populated by an initcontainer I think we are good for live migration? I suppose the warning can be more specific to say initcontainer rather than sidecar though

**Comment by fossedihelm** [comment]
@mhenriks Can you rebase please? Thanks

**Comment by mhenriks** [comment]
> And last, (sorry if you did it, I didn't find it) I think it should block the live migration with empty dirs, I don't think a warning in the docs is enough.
> 
> The guest kernel doesn't handle well that files disappears while in the icache/dcache, I plan to work on it this year to improve the migration reliability.

@germag can you confirm that we are good with emptydirs since they are read-only?

File: cmd/test-helpers/BUILD.bazel (hunk 1 above)
Comment (0xFelix) [reply cmd/test-helpers/BUILD.bazel:22]: IIUC quay would only be required if you intend to publish this image. If you want it to be available in local / testing environments only (`cluster-sync`), it should be okay to only define it in-tree. 

Have a look at `hack/bazel-push-images.sh` and `hack/virt-template/push-images.sh`.

If you want to push it to quay then @dhiller would need to add another repo on quay for it and you would have to add it to `hack/bazel-push-images.sh`

Comment (mhenriks) [reply cmd/test-helpers/BUILD.bazel:22]: I don't think there's any differentiation between what we push "locally" and to quay. Had to add the image to `hack/bazel-push-images.sh` in order for `cluster-sync` to push the image

Basically I'm not sure if the kubevirt org on quay is configured to allow pushing of new images or if all possible images have to be set up before first push

Comment (0xFelix) [reply cmd/test-helpers/BUILD.bazel:22]: You can differentiate if you don't put the image into `hack/bazel-push-images.sh` and use a separate pushing script instead (virt-template works this way).

@dhiller set up the repo first everytime I needed a new image in the kubevirt org on quay

Comment (mhenriks) [reply cmd/test-helpers/BUILD.bazel:22]: @0xFelix I understand it's possible, but I'm not sure it makes sense to handle this image differently than other test specific images like `vm-killer`, `disks-images-provider`, `example-hook-sidecar`

Comment (0xFelix) [reply cmd/test-helpers/BUILD.bazel:22]: I'm sure @dhiller will be happy to create a new repo. :)

**Comment by mhenriks** [comment]
Confirmed working on aws with this VM spec:

```yaml
kind: VirtualMachine
metadata:
  name: testvm
  namespace: default
spec:
  runStrategy: Always
  template:
    metadata:
      labels:
        app: testvm
    spec:
      domain:
        resources:
          requests:
            memory: 512Mi
        devices:
          filesystems:
            - name: serviceaccount-fs
              virtiofs: {}
            - name: aws-token
              virtiofs: {}
          disks:
            - name: rootdisk
              disk:
                bus: virtio
          interfaces:
            - name: default
              masquerade: {}
        machine:
          type: q35
      networks:
        - name: default
          pod: {}
      volumes:
        - name: rootdisk
          containerDisk:
            image: quay.io/kubevirt/fedora-with-test-tooling:v20240717-a087e7e
        - name: serviceaccount-fs
          serviceAccount:
            serviceAccountName: my-service-account
        - name: aws-token
          containerPath:
            path: /var/run/secrets/eks.amazonaws.com/serviceaccount
```

```bash
➜  ~ virtctl console testvm
Successfully connected to testvm console. The escape sequence is ^]

[fedora@testvm ~]$
[fedora@testvm ~]$ sudo mkdir /mnt/iam-token
[fedora@testvm ~]$ sudo mount -t virtiofs aws-token /mnt/iam-token
[  303.760316] SELinux: (dev virtiofs, type virtiofs) has no security xattr handler
[  303.762057] SELinux: (dev virtiofs, type virtiofs) falling back to genfs
[fedora@testvm ~]$
[fedora@testvm ~]$ sudo cat /mnt/iam-token/token
eyJhbGciOiJSUzI1NiIsImtpZCI6ImpmdDZ0a0RMRkZ3Qi14UEl0Mldxa0RYVWptM0NRSU9LQ2NZdlNhbnd1NDQifQ.eyJ
<CUT DATA>
4MP2M7K7T9tE5BlIas42jSQBN_dAtb5iIh3sY--vffCQuHFNAiKu0hnwyYfLp7DmIv93jfs2YxZm8bHCEiaC1EHsvitB8qbWX3jmg8GcPf4auRCLZ3DcRbe44rvOeEQ[fedora@testvm ~]$
[fedora@testvm ~]$
[fedora@testvm ~]$
```

File: cmd/test-helpers/BUILD.bazel (hunk 1 above)
Comment (akalenyu) [reply cmd/test-helpers/BUILD.bazel:22]: was going to suggest https://kubernetes.io/docs/reference/access-authn-authz/mutating-admission-policy/ since it's beta in 1.34 but openshift still doesn't have it enabled by default. your call

File: tests/operator/operator.go [hunk 2]
```diff
@@ -1836,6 +1836,42 @@ var _ = Describe("[sig-operator]Operator", Serial, decorators.SigOperator, func(
 		})
 	})
 
+	Context("with ContainerPathVolumes feature gate toggled", func() {
```
Comment (akalenyu) [inline tests/operator/operator.go:1839]: should we unit test this instead? not worth the control plane reroll

File: cmd/test-helpers/BUILD.bazel (hunk 1 above)
Comment (mhenriks) [reply cmd/test-helpers/BUILD.bazel:22]: yeah was mostly concerned about version compatibility. 1.34 is only about 6 months old. But may be nice to replace with later

Less of a concern was that MutatingWebhook is closer to real world testing

File: tests/virtiofs/containerpath.go [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+		const (
+			containerPathFilesystemName = "nonexistent-path"
+			nonExistentPath             = "/this/path/does/not/exist"
+		)
+
+		It("Should set Synchronized=False with MissingVirtiofsContainers reason", func() {
+			virtClient := kubevirt.Client()
+
+			By("Creating a VMI with ContainerPath pointing to non-existent path")
+			vmi := libvmifact.NewFedora(
```
Comment (akalenyu) [inline tests/virtiofs/containerpath.go]: we could use alpine/guestless here right? fedora requirements are a bit on the higher side

File: tests/virtiofs/containerpath.go [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+				&expect.BSnd{S: "echo $?\n"},
+				&expect.BExp{R: console.RetValue("0")},
+				// Read the test file that was written from the pod
+				&expect.BSnd{S: fmt.Sprintf("cat /mnt/%s\n", testFileName)},
+				&expect.BExp{R: testContent},
+			}, 200)).To(Succeed())
+		})
+	})
+
+	Context("With webhook-injected ConfigMap volume and migration", Serial, func() {
```
Comment (akalenyu) [inline tests/virtiofs/containerpath.go]: Why is the test serial? each webhook targets a unique test namespace? what am I missing

File: pkg/virt-operator/resource/generate/components/webhooks.go [hunk 5]
```diff
@@ -984,3 +984,65 @@ const MigrationPolicyCreateValidatePath = "/migration-policy-validate-create"
...
+const VirtLauncherPodMutatePath = "/virt-launcher-pod-mutate"
+
+const VirtLauncherPodMutatingWebhookName = "virt-launcher-pod-mutator"
+
+func NewVirtLauncherPodMutatingWebhookConfiguration(installNamespace string) *admissionregistrationv1.MutatingWebhookConfiguration {
+	path := VirtLauncherPodMutatePath
+	// Use Ignore so we don't block pod creation if virt-api is unavailable
+	failurePolicy := admissionregistrationv1.Ignore
+	// Reinvoke if other mutators make changes, ensuring we run after external mutators
+	reinvocationPolicy := admissionregistrationv1.IfNeededReinvocationPolicy
```
Comment (akalenyu) [inline pkg/virt-operator/resource/generate/components/webhooks.go:997]: I think we could benefit from unit tests that assert these 2

File: pkg/virt-api/webhooks/mutating-webhook/mutators/virt-launcher-pod-mutator.go [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+	"kubevirt.io/kubevirt/pkg/pointer"
+	"kubevirt.io/kubevirt/pkg/util"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+	"kubevirt.io/kubevirt/pkg/virtiofs"
+)
+
+const (
+	computeContainerName = "compute"
```
Comment (akalenyu) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/virt-launcher-pod-mutator.go:43]: we really don't have a variable for that huh

**Review by akalenyu — COMMENTED** [review]
looks good!

File: tests/operator/operator.go (hunk 2 above)
Comment (mhenriks) [reply tests/operator/operator.go:1839]: We do a similar test for toggling vmexport featuregate I think it's nice to have

File: tests/virtiofs/containerpath.go (hunk 3 above)
Comment (mhenriks) [reply tests/virtiofs/containerpath.go]: done

File: tests/virtiofs/containerpath.go (hunk 4 above)
Comment (mhenriks) [reply tests/virtiofs/containerpath.go]: technically changes cluster state, figured better safe than sorry

File: pkg/virt-operator/resource/generate/components/webhooks.go (hunk 5 above)
Comment (mhenriks) [reply pkg/virt-operator/resource/generate/components/webhooks.go:997]: shocked to see there were no unit tests so added some

File: pkg/virt-api/webhooks/mutating-webhook/mutators/virt-launcher-pod-mutator.go (hunk 6 above)
Comment (mhenriks) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/virt-launcher-pod-mutator.go:43]: I don't think so

File: tests/virtiofs/containerpath.go (hunk 4 above)
Comment (mhenriks) [reply tests/virtiofs/containerpath.go]: screw it yolo

**Comment by germag** [comment]
> > And last, (sorry if you did it, I didn't find it) I think it should block the live migration with empty dirs, I don't think a warning in the docs is enough.
> > The guest kernel doesn't handle well that files disappears while in the icache/dcache, I plan to work on it this year to improve the migration reliability.
> 
> @germag can you confirm that we are good with emptydirs since they are read-only?

Yes, but I think it should be documented that the user must keep the migration-source and target dir in sync

File: tests/virtiofs/containerpath.go (hunk 4 above)
Comment (akalenyu) [reply tests/virtiofs/containerpath.go]: maybe use generateName for the webhook? other than that what else are you worried about?

File: pkg/virt-operator/kubevirt_test.go [hunk 7]
```diff
@@ -3746,6 +3929,7 @@ var _ = Describe("KubeVirt Operator", func() {
 				},
 			}
 			enableTemplateFeatureGate(kv)
+			enableContainerPathVolumesFeatureGate(kv)
```
Comment (akalenyu) [inline pkg/virt-operator/kubevirt_test.go:3932]: @0xFelix are we sure this approach of adding the gate to all unit tests is fine? seems kind of odd tbh

Comment (0xFelix) [reply pkg/virt-operator/kubevirt_test.go:3932]: Not adding the FG to all tests would mean we would have to duplicate all existing tests to also enable the feature gate or have only a small subset of tests with this FG enabled. Sounds odd to me as well. 

By adding them unconditionally we know pretty early when stuff breaks?

Or we could refactor all test cases to be tables... but the more FGs we have, the more it would explode complexity.

Comment (0xFelix) [reply pkg/virt-operator/kubevirt_test.go:3932]: I'm open to refactorings though.

File: cmd/test-helpers/BUILD.bazel (hunk 1 above)
Comment (dhiller) [reply cmd/test-helpers/BUILD.bazel:22]: apologies for missing this, https://quay.io/repository/kubevirt/test-helpers created

**Review by akalenyu — APPROVED** [review]

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 8]
```diff
@@ -284,28 +284,63 @@ func validateDownwardMetrics(field *k8sfield.Path, spec *v1.VirtualMachineInstan
...
+			}
+			if !matchingFilesystemExists {
+				causes = append(causes, metav1.StatusCause{
+					Type:    metav1.CauseTypeFieldValueInvalid,
+					Message: fmt.Sprintf("ContainerPath volume '%s' must have a matching filesystem defined in spec.domain.devices.filesystems", volume.Name),
+					Field:   field.Child("domain", "volumes").Index(idx).String(),
+				})
+			}
+			// Block reserved paths used by KubeVirt internally
+			reservedPrefixes := []string{"/var/run/kubevirt", "/var/run/libvirt"}
```
Comment (xpivarc) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:332]: Should we block equivalent `/run/*` ?

File: pkg/virt-controller/watch/vmi/lifecycle.go [hunk 9]
```diff
@@ -136,6 +137,14 @@ func (c *Controller) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod, da
 			return common.NewSyncError(fmt.Errorf("PVC pending"), controller.BackendStorageNotReadyReason), pod
 		}
 
+		// Block pod creation if VMI uses ContainerPath volumes but feature gate is disabled
+		containerPathVolumes := virtiofs.GetContainerPathVolumesWithFilesystems(vmi)
```
Comment (xpivarc) [inline pkg/virt-controller/watch/vmi/lifecycle.go:141]: We usually do this in the admiter/webhooks, why is this here?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go (hunk 8 above)
Comment (mhenriks) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:332]: sure!

File: pkg/virt-controller/watch/vmi/lifecycle.go (hunk 9 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/vmi/lifecycle.go:141]: This is mostly to handle the case where the featuregate is enabled and disabled later. Can handle CREATE in admitter but have to be smart about handling UPDATE to handle the case where the featuregate is enabled. Can add that for beta

Comment (xpivarc) [reply pkg/virt-controller/watch/vmi/lifecycle.go:141]: >Can add that for beta

Please do, that would be great!

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go (hunk 8 above)
Comment (xpivarc) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:332]: Thanks, let's track in VEP issue
